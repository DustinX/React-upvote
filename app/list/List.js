import React, {useState} from 'react';
import Product from './Product';


const List = () => {
  
  const [products, setProducts] = useState(window.Seed.products);
  
  return (
    <div className="list_page">
      <h1 className='header'>Popular Products</h1>
      <hr></hr>
      <ul>{ products
              .map( product => <Product key={product.id} product={product} />)} 
      </ul>
    </div>
  )
}

export default List;
