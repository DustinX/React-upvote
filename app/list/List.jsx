import React, { useState } from 'react';
import Product from './Product';


const List = () => {
  // eslint-disable-next-line no-undef
  const [products, setProducts] = useState(window.Seed.products);

  return (
    <div className="list_page">
      <h1 className="header">Popular Products</h1>
      <hr />
      <ul>
        {products
          .sort((a, b) => b.votes - a.votes)
          .map((product) => (
            <Product
              key={product.id}
              product={product}
              products={products}
              setProducts={setProducts}
            />
          ))}
      </ul>
    </div>
  );
};

export default List;
