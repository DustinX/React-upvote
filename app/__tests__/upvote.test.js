import React from 'react';
import {fireEvent, render} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer';
import List from '../list/List';



// initialization
window.Seed = {};

function initializeProducts () {
  
  window.Seed.products = [
  {
    id: 1,
    title: 'Yellow Pail',
    description: 'On-demand sand castle construction expertise.',
    url: '#',
    votes: 10,
    submitterAvatarUrl: 'images/avatars/daniel.jpg',
    productImageUrl: 'images/products/image-aqua.png',
  },
  {
    id: 2,
    title: 'Supermajority: The Fantasy Congress League',
    description: 'Earn points when your favorite politicians pass legislation.',
    url: '#',
    votes: 9,
    submitterAvatarUrl: 'images/avatars/kristy.png',
    productImageUrl: 'images/products/image-rose.png',
  }]
  
}

beforeEach(() => {
  initializeProducts();
});

// tests
it('renders correctly', () => {
  const tree = renderer
    .create(<List/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Upvote count increases after click', () => {
  
  const {getByText, queryAllByTestId} = render(
    <List />
  );

  expect(getByText('10')).toBeTruthy();
  
  fireEvent.click(queryAllByTestId(/vote-button/i)[0]);
  
  expect(getByText('11')).toBeTruthy();
});



it('switches order correctly', () => {
  
  const {queryAllByTestId} = render(
    <List />
  );
  
  expect( queryAllByTestId(/product/i)[0] ).toHaveTextContent(/Yellow/)

  //upvote twice to make second item become first
  fireEvent.click(queryAllByTestId(/vote-button/i)[1]);
  fireEvent.click(queryAllByTestId(/vote-button/i)[1]);
  
  expect( queryAllByTestId(/product/i)[0] ).toHaveTextContent(/Supermajority/)
  
});
