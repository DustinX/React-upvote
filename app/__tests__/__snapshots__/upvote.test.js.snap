// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="list_page"
>
  <h1
    className="header"
  >
    Popular Products
  </h1>
  <hr />
  <ul>
    <li
      className="product"
      data-testid="product"
    >
      <img
        className="image"
        src="images/products/image-aqua.png"
      />
      <img
        className="arrow"
        data-testid="vote-button"
        onClick={[Function]}
        src="images/ui/arrow.png"
      />
      <p
        className="votes"
        data-testid="votes"
      >
        10
      </p>
      <a
        className="title"
        href="#"
      >
        Yellow Pail
      </a>
      <p
        className="description"
      >
        On-demand sand castle construction expertise.
      </p>
      <p
        className="submitted"
      >
        Submitted by:
      </p>
      <img
        className="submitter"
        src="images/avatars/daniel.jpg"
      />
    </li>
    <li
      className="product"
      data-testid="product"
    >
      <img
        className="image"
        src="images/products/image-rose.png"
      />
      <img
        className="arrow"
        data-testid="vote-button"
        onClick={[Function]}
        src="images/ui/arrow.png"
      />
      <p
        className="votes"
        data-testid="votes"
      >
        9
      </p>
      <a
        className="title"
        href="#"
      >
        Supermajority: The Fantasy Congress League
      </a>
      <p
        className="description"
      >
        Earn points when your favorite politicians pass legislation.
      </p>
      <p
        className="submitted"
      >
        Submitted by:
      </p>
      <img
        className="submitter"
        src="images/avatars/kristy.png"
      />
    </li>
  </ul>
</div>
`;
